diff -drup org/gcc-4.6.1/configure gcc-4.6.1/configure
--- org/gcc-4.6.1/configure	2011-03-16 19:27:36.000000000 +0100
+++ gcc-4.6.1/configure	2011-08-07 11:17:15.000000000 +0200
@@ -2731,6 +2731,9 @@ target_libraries="target-libgcc \
 #
 target_tools="target-examples target-groff target-gperf target-rda"
 
+target_libraries=""
+target_tools=""
+
 ################################################################################
 
 ## All tools belong in one of the four categories, and are assigned above
diff -drup org/gcc-4.6.1/gcc/config/mips/mips.c gcc-4.6.1/gcc/config/mips/mips.c
--- org/gcc-4.6.1/gcc/config/mips/mips.c	2011-05-29 19:48:14.000000000 +0200
+++ gcc-4.6.1/gcc/config/mips/mips.c	2011-08-07 20:11:26.000000000 +0200
@@ -706,6 +706,7 @@ static const struct mips_cpu_info mips_c
   /* ST Loongson 2E/2F processors.  */
   { "loongson2e", PROCESSOR_LOONGSON_2E, 3, PTF_AVOID_BRANCHLIKELY },
   { "loongson2f", PROCESSOR_LOONGSON_2F, 3, PTF_AVOID_BRANCHLIKELY },
+  { "f32c", PROCESSOR_F32C, 3, 0 },
 
   /* MIPS IV processors. */
   { "r8000", PROCESSOR_R8000, 4, 0 },
@@ -1183,7 +1184,16 @@ static const struct mips_rtx_cost_data
     COSTS_N_INSNS (72),           /* int_div_di */
 		     1,           /* branch_cost */
 		     4            /* memory_latency */
-  }
+  },
+  { /* F32C */
+    SOFT_FP_COSTS,
+    COSTS_N_INSNS (3),            /* int_mult_si */
+    COSTS_N_INSNS (3),            /* int_mult_di */
+    COSTS_N_INSNS (36),           /* int_div_si */
+    COSTS_N_INSNS (68),           /* int_div_di */
+		     1,           /* branch_cost */
+		     3            /* memory_latency */
+  },
 };
 
 static rtx mips_find_pic_call_symbol (rtx, rtx, bool);
diff -drup org/gcc-4.6.1/gcc/config/mips/mips.h gcc-4.6.1/gcc/config/mips/mips.h
--- org/gcc-4.6.1/gcc/config/mips/mips.h	2011-03-08 21:51:11.000000000 +0100
+++ gcc-4.6.1/gcc/config/mips/mips.h	2011-08-07 20:06:20.000000000 +0200
@@ -219,6 +219,7 @@ enum mips_code_readable_setting {
 #define TARGET_LOONGSON_2F          (mips_arch == PROCESSOR_LOONGSON_2F)
 #define TARGET_LOONGSON_2EF         (TARGET_LOONGSON_2E || TARGET_LOONGSON_2F)
 #define TARGET_LOONGSON_3A          (mips_arch == PROCESSOR_LOONGSON_3A)
+#define TARGET_F32C                 (mips_arch == PROCESSOR_F32C)
 #define TARGET_MIPS3900             (mips_arch == PROCESSOR_R3900)
 #define TARGET_MIPS4000             (mips_arch == PROCESSOR_R4000)
 #define TARGET_MIPS4120             (mips_arch == PROCESSOR_R4120)
@@ -244,6 +245,7 @@ enum mips_code_readable_setting {
 #define TUNE_LOONGSON_2EF           (mips_tune == PROCESSOR_LOONGSON_2E	\
 				     || mips_tune == PROCESSOR_LOONGSON_2F)
 #define TUNE_LOONGSON_3A            (mips_tune == PROCESSOR_LOONGSON_3A)
+#define TUNE_F32C                   (mips_tune == PROCESSOR_F32C)
 #define TUNE_MIPS3000               (mips_tune == PROCESSOR_R3000)
 #define TUNE_MIPS3900               (mips_tune == PROCESSOR_R3900)
 #define TUNE_MIPS4000               (mips_tune == PROCESSOR_R4000)
@@ -712,10 +714,11 @@ enum mips_code_readable_setting {
   "%{" MIPS_ISA_LEVEL_OPTION_SPEC ":;: \
      %{march=mips1|march=r2000|march=r3000|march=r3900:-mips1} \
      %{march=mips2|march=r6000:-mips2} \
-     %{march=mips3|march=r4*|march=vr4*|march=orion|march=loongson2*:-mips3} \
+     %{march=mips3|march=r4*|march=vr4*|march=orion|march=loongson2* \
+       |march=f32c: -mips3} \
      %{march=mips4|march=r8000|march=vr5*|march=rm7000|march=rm9000 \
        |march=r10000|march=r12000|march=r14000|march=r16000:-mips4} \
-     %{march=mips32|march=4kc|march=4km|march=4kp|march=4ksc:-mips32} \
+     %{march=mips32|march=4kc|march=4km|march=4kp|march=4ksc: -mips32} \
      %{march=mips32r2|march=m4k|march=4ke*|march=4ksd|march=24k* \
        |march=34k*|march=74k*|march=1004k*: -mips32r2} \
      %{march=mips64|march=5k*|march=20k*|march=sb1*|march=sr71000 \
@@ -731,7 +734,7 @@ enum mips_code_readable_setting {
   "%{mhard-float|msoft-float|march=mips*:; \
      march=vr41*|march=m4k|march=4k*|march=24kc|march=24kec \
      |march=34kc|march=74kc|march=1004kc|march=5kc \
-     |march=octeon|march=xlr: -msoft-float;		  \
+     |march=octeon|march=xlr|march=f32c: -msoft-float;		  \
      march=*: -mhard-float}"
 
 /* A spec condition that matches 32-bit options.  It only works if
@@ -847,7 +850,7 @@ enum mips_code_readable_setting {
 
 /* ISA has the integer conditional move instructions introduced in mips4 and
    ST Loongson 2E/2F.  */
-#define ISA_HAS_CONDMOVE        (ISA_HAS_FP_CONDMOVE || TARGET_LOONGSON_2EF)
+#define ISA_HAS_CONDMOVE        (ISA_HAS_FP_CONDMOVE || TARGET_LOONGSON_2EF || TARGET_F32C)
 
 /* ISA has LDC1 and SDC1.  */
 #define ISA_HAS_LDC1_SDC1	(!ISA_MIPS1 && !TARGET_MIPS16)
diff -drup org/gcc-4.6.1/gcc/config/mips/mips.md gcc-4.6.1/gcc/config/mips/mips.md
--- org/gcc-4.6.1/gcc/config/mips/mips.md	2011-03-03 22:56:58.000000000 +0100
+++ gcc-4.6.1/gcc/config/mips/mips.md	2011-08-07 20:46:52.000000000 +0200
@@ -63,6 +63,7 @@
   sb1a
   sr71000
   xlr
+  f32c
 ])
 
 (define_c_enum "unspec" [
@@ -913,6 +914,7 @@
 (include "sb1.md")
 (include "sr71k.md")
 (include "xlr.md")
+(include "f32c.md")
 (include "generic.md")
 
 ;;
@@ -1323,7 +1325,7 @@
   [(set (match_operand:GPR 0 "register_operand")
 	(mult:GPR (match_operand:GPR 1 "register_operand")
 		  (match_operand:GPR 2 "register_operand")))]
-  ""
+  "!TARGET_SWMUL"
 {
   if (TARGET_LOONGSON_2EF || TARGET_LOONGSON_3A)
     emit_insn (gen_mul<mode>3_mul3_loongson (operands[0], operands[1],
@@ -1731,7 +1733,7 @@
   [(set (match_operand:DI 0 "register_operand")
 	(mult:DI (any_extend:DI (match_operand:SI 1 "register_operand"))
 		 (any_extend:DI (match_operand:SI 2 "register_operand"))))]
-  "mips_mulsidi3_gen_fn (<CODE>) != NULL"
+  "!TARGET_SWMUL && mips_mulsidi3_gen_fn (<CODE>) != NULL"
 {
   mulsidi3_gen_fn fn = mips_mulsidi3_gen_fn (<CODE>);
   emit_insn (fn (operands[0], operands[1], operands[2]));
@@ -1906,7 +1908,7 @@
 	  (mult:DI (any_extend:DI (match_operand:SI 1 "register_operand"))
 		   (any_extend:DI (match_operand:SI 2 "register_operand")))
 	  (const_int 32))))]
-  ""
+  "!TARGET_SWMUL"
 {
   if (ISA_HAS_MULHI)
     emit_insn (gen_<su>mulsi3_highpart_mulhi_internal (operands[0],
@@ -1926,7 +1928,7 @@
 		   (any_extend:DI (match_operand:SI 2 "register_operand" "d")))
 	  (const_int 32))))
    (clobber (match_scratch:SI 3 "=l"))]
-  "!ISA_HAS_MULHI"
+  "!TARGET_SWMUL && !ISA_HAS_MULHI"
   { return TARGET_FIX_R4000 ? "mult<u>\t%1,%2\n\tmfhi\t%0" : "#"; }
   "&& reload_completed && !TARGET_FIX_R4000"
   [(const_int 0)]
@@ -2250,7 +2252,9 @@
 		  (match_operand:ANYF 2 "register_operand")))]
   "<divide_condition>"
 {
-  if (const_1_operand (operands[1], <MODE>mode))
+  if (TARGET_SWDIV)
+    FAIL;
+  else if (const_1_operand (operands[1], <MODE>mode))
     if (!(<recip_condition> && flag_unsafe_math_optimizations))
       operands[1] = force_reg (<MODE>mode, operands[1]);
 })
@@ -2312,7 +2316,7 @@
    (set (match_operand:GPR 3 "register_operand" "=d")
 	(mod:GPR (match_dup 1)
 		 (match_dup 2)))]
-  "!TARGET_FIX_VR4120"
+  "!TARGET_SWDIV && !TARGET_FIX_VR4120"
   "#"
   "&& reload_completed"
   [(const_int 0)]
@@ -2344,7 +2348,7 @@
    (set (match_operand:GPR 3 "register_operand" "=d")
 	(umod:GPR (match_dup 1)
 		  (match_dup 2)))]
-  ""
+  "!TARGET_SWDIV"
   "#"
   "reload_completed"
   [(const_int 0)]
@@ -2375,7 +2379,7 @@
 	  [(any_div:GPR (match_operand:GPR 1 "register_operand" "d")
 			(match_operand:GPR 2 "register_operand" "d"))]
 	  UNSPEC_SET_HILO))]
-  ""
+  "!TARGET_SWDIV"
   { return mips_output_division ("<GPR:d>div<u>\t%.,%1,%2", operands); }
   [(set_attr "type" "idiv")
    (set_attr "mode" "<GPR:MODE>")])
@@ -6301,7 +6305,7 @@
 		 (const_int 0)])
 	 (match_operand:GPR 2 "reg_or_0_operand" "dJ,0")
 	 (match_operand:GPR 3 "reg_or_0_operand" "0,dJ")))]
-  "ISA_HAS_CONDMOVE"
+  "ISA_HAS_CONDMOVE && !TARGET_F32C"
   "@
     mov%T4\t%0,%z2,%1
     mov%t4\t%0,%z3,%1"
diff -drup org/gcc-4.6.1/gcc/config/mips/mips.opt gcc-4.6.1/gcc/config/mips/mips.opt
--- org/gcc-4.6.1/gcc/config/mips/mips.opt	2011-02-17 02:59:04.000000000 +0100
+++ gcc-4.6.1/gcc/config/mips/mips.opt	2011-08-07 11:11:03.000000000 +0200
@@ -76,6 +76,14 @@ mdouble-float
 Target Report RejectNegative InverseMask(SINGLE_FLOAT, DOUBLE_FLOAT)
 Allow hardware floating-point instructions to cover both 32-bit and 64-bit operations
 
+msoft-mul
+Target Report Mask(SWMUL)
+Do not use integer mul instructions
+
+msoft-div
+Target Report Mask(SWDIV)
+Do not use integer div instructions
+
 mdsp
 Target Report Mask(DSP)
 Use MIPS-DSP instructions
diff -drup org/gcc-4.6.1/gcc/doc/g++.1 gcc-4.6.1/gcc/doc/g++.1
--- org/gcc-4.6.1/gcc/doc/g++.1	2011-06-27 12:49:31.000000000 +0200
+++ gcc-4.6.1/gcc/doc/g++.1	2011-08-07 11:11:03.000000000 +0200
@@ -13122,6 +13122,7 @@ The processor names are:
 \&\fB34kc\fR, \fB34kf2_1\fR, \fB34kf1_1\fR,
 \&\fB74kc\fR, \fB74kf2_1\fR, \fB74kf1_1\fR, \fB74kf3_2\fR,
 \&\fB1004kc\fR, \fB1004kf2_1\fR, \fB1004kf1_1\fR,
+\&\fBf32c\fR,
 \&\fBloongson2e\fR, \fBloongson2f\fR, \fBloongson3a\fR,
 \&\fBm4k\fR,
 \&\fBocteon\fR,
diff -drup org/gcc-4.6.1/gcc/doc/gcc.1 gcc-4.6.1/gcc/doc/gcc.1
--- org/gcc-4.6.1/gcc/doc/gcc.1	2011-06-27 12:49:31.000000000 +0200
+++ gcc-4.6.1/gcc/doc/gcc.1	2011-08-07 11:11:03.000000000 +0200
@@ -13122,6 +13122,7 @@ The processor names are:
 \&\fB34kc\fR, \fB34kf2_1\fR, \fB34kf1_1\fR,
 \&\fB74kc\fR, \fB74kf2_1\fR, \fB74kf1_1\fR, \fB74kf3_2\fR,
 \&\fB1004kc\fR, \fB1004kf2_1\fR, \fB1004kf1_1\fR,
+\&\fBf32c\fR,
 \&\fBloongson2e\fR, \fBloongson2f\fR, \fBloongson3a\fR,
 \&\fBm4k\fR,
 \&\fBocteon\fR,
