--- project/project_project_sram.svf.orig	2016-09-03 16:33:00.943302205 +0200
+++ project/project_project_sram.svf	2016-09-03 16:37:54.117492317 +0200
@@ -45,10 +45,10 @@ STATE	IDLE;
 ! Check the IDCODE
 
 ! Shift in IDCODE(0x16) instruction
-SIR	8	TDI  (16);
-SDR	32	TDI  (FFFFFFFF)
-		TDO  (01015043)
-		MASK (FFFFFFFF);
+!SIR	8	TDI  (16);
+!SDR	32	TDI  (FFFFFFFF)
+!		TDO  (01015043)
+!		MASK (FFFFFFFF);
 
 
 ! Program Bscan register
@@ -96166,7 +96166,8 @@ RUNTEST	IDLE	100 TCK	1,00E-03 SEC;
 ! Verify Status Register
 
 ! Shift in LSCC READ STATUS(0x53) instruction
-SIR	8	TDI  (53);
-SDR	32	TDI  (FFFFFFFF)
-		TDO  (00020000)
-		MASK (00060007);
+!SIR	8	TDI  (53);
+!SDR	32	TDI  (FFFFFFFF)
+!		TDO  (00020000)
+!		MASK (00060007);
+!
\ Nema novog retka na kraju datoteke
